### ISSUE TITLE
Proposal: a `toISOString` function for ease of adoption

### ISSUE THREAD

**Issue by dgreene1** [js-joda/js-joda#400]
First let me say that I love this library and I often evangelize that more people should use it.

But one of the "barriers to entry" in my opinion is the fact that it's difficult to serialize the various types to string before I send them across the wire to my back-end code.

## Proposal

Each type would contain a `.toISOString` function.

```ts
LocalDate.now().toISOString(); // "2020-04-20"
LocalDateTime.now().toISOString(); // "2020-04-20T15:59:50.985"
ZonedDateTime.now().toISOString(); // "2020-04-20T15:59:50.985Z"
// etc.
```

## disclaimer

Yes, I know that this is just syntactic sugar, however it would be a massive improvement for explaining to developers how to use this.

I recognize that some might see it as an unsafe "escape hatch" where you could leave the strict types to get to an unsafe string but I don't believe that is a true concern due to the strictness of js-joda's parse functions. Like it would be challenging if not impossible to convert between too different types.

```ts
// This is still not possible, which means that this is a safe proposal
ZonedDateTime.parse(LocalDate.now().toISOString()) // would throw an error
```

**Comment by dgreene1** [comment]
I believe @pithu and/or @phueper would be able to provide the best input on (a) why this function hasn't been included so far and (b) if a PR would be accepted that includes it.

I would be happy to submit this PR with a healthy amount of unit test coverage. :)

**Comment by pithu** [comment]
Hi @dgreene1 

Why not simply using the existing `toString()` method instead?

**Comment by dgreene1** [comment]
Because `.toString` doesn't always output a valid ISO string. It only does for certain types.

```ts
ZonedDateTime.now().toString(); // "2020-04-20T12:00:31.476-04:00[America/New_York]"
```
But with the proposed method (`.toISOString`), it would produce a valid ISO string.

**Comment by pithu** [comment]
I don't get your point. `2020-04-20T12:00:31.476-04:00[America/New_York]` is a valid ISO8601 string. Actually ISO8601 allows a lot of format variants. ThreeTen (and therefore js-joda) supports only a subset.

And it should be possible to serialize/ deserialize with `.parse(...)` and `.toString()` all js-joda types. If not, its probably a bug.

**Comment by dgreene1** [comment]
So the most typical use case for `.toISOString` is to turn it into a string so you can send it to an API. I've included a C# REPL link so that you can see how the format that ZonedDateTime's `.toString` provides isn't accepted by most date libraries. They're expecting the UTC ISO format.

https://dotnetfiddle.net/9sx3bS

From the research I have done, I do not believe that the format of ZonedDateTime's `.toString` function is part of ISO 8601. Perhaps and extension, but not the standard that most libraries use.

As evidence of this, check out the [Java docs](https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeFormatter.html#ISO_ZONED_DATE_TIME) that clearly state that the format is not part of the ISO-8601 standard:

> public static final DateTimeFormatter ISO_ZONED_DATE_TIME
> 
> The ISO-like date-time formatter that formats or parses a date-time with offset and zone, such as '2011-12-03T10:15:30+01:00[Europe/Paris]'.
> 
> This returns an immutable formatter capable of formatting and parsing a format that extends the ISO-8601 extended offset date-time format to add the time-zone. The section in square brackets is not part of the ISO-8601 standard. 

Here's another link to [another date library](https://github.com/moment/moment/issues/4159#issuecomment-329437605) that doesn't believe that this part of the ISO standard:
> Gonna need a better citation that this is part of ISO 8601. I don't think it is; I think Java just added it because it's nice. 

And my last reason why we should do this is because it's what JavaScript users expect.
```ts
const zdt = window.jsJoda.ZonedDateTime.now();
zdt.toString();
// "2020-04-23T11:28:41.735-04:00[America/New_York]"
zdt.toISOString();
"2020-04-23T15:28:41.735Z"
```

I want to be clear that I am aware that this is just syntactic sugar over `ZonedDateTime.now().toInstant().toString()`; however, given the advantages it should be something we consider highly:
* better interop with most date libraries (I've tested that the current format breaks with C#, MomentJS, and python)
* more familiar DX (developer experience)
* it will help me to convince more people (including 50+ developers at my company and 1000s of people on my blog) that they should use js-joda

**Comment by pithu** [comment]
Ok, i got your point and yes, timezones like `[America/New_York]` are not specified in ISO8601. 

Nevertheless ISO8601 allows a lot of different formats, therefore the output of `toISOString()` couldn't be clearly specified. 

Furthermore it sounds like syntactic sugar and we have the agreement to stick to the threeten API, i am not convinced to change that. Sorry.

**Comment by dgreene1** [comment]
> Furthermore it sounds like syntactic sugar and we have the agreement to stick to the threeten API, i am not convinced to change that. Sorry.

I would really love for you to reconsider this position as it only prevents my ability to evangelize the use of this wonderful library to other developers. And if I can't evangelize this library then I have no way of helping teammates avoid bugs caused by domain issues such as the differences between a day and an instant that plagues most MomentJS users.

Perhaps if we understood the rationale for the strict adherence threeten API it would be more palatable for future users to accept the lack of this pivitol JS method (i.e. `.toISODate`). But even then, I remain unable to confidently say to other developers "yes, it's easy to use js-joda to send dates over HTTP" due to the difficulty they might have in trying to figure out what the proper set of methods are that they would need to use to get a valid ISO-8601 string.

If a full stack developer is using TypeScript in the UI and C# in the backend, they will get a deserialization error even if they are using NodaTime (which is very similar to the Threeten API) simply because ZonedDateTime's serialization method doesn't output a string that NodaTime can successfully parse.

Please understand that my only intention here is to help more people to find, enjoy, and to share this wonderful library. :) And regardless of your decision I sincerely appreciate the time that you and your peers spend on maintaining this library.

**Comment by phueper** [comment]
Hi @dgreene1,

i agree with @pithu here. We try to adhere to the threeten API as much as possible to make it easy to transition between the java.time or other threeten libraries usage and js-joda usage. So i don't think this extension makes this transition better.
That said, i also have a problem with your proposal from the perspective of the domain design of the API. I believe the beauty of the joda/threeten API design is in the clearness of the expressions we can use. If we would have a `toISOString` function in `ZonedDateTime`, in my opinion it hints to  a user of the library, that this String can be used to recreate the same ZonedDateTime from it, but that is not the case since we are losing the Zone Information. If we want to recreate the same ZonedDateTime, we somehow need to transport the information in some other way. However, if it is clear, that the transported Object in the String is "only" an Instant by using `zdt.toInstant().toString()` the developer has to make the clear decision to drop the zone information (by using `.toInstant` and can decide wether that is ok for him or if he needs to pass the zone information in some other way). 
Besides, i really don't see the big advantage of using `toISOString()` over `.toInstant().toString()`, it's just a couple of characters and the latter version is imho **much** clearer in what is happening.

Regards, Pattrick

**Comment by dgreene1** [comment]
Thank you @phueper. I think your answer together with @pithu‘s give future developers who visit this thread enough information to feel good about the lack of a`.toISOString` function. So I really appreciate both of your input.

I suppose my hangout is only in relation to ZonedDateTime’s toString method since it’s the only toString method in this library that doesn’t output a standard ISO-8601 format.

So maybe the solution for me is to communicate to fellow developers that they should perhaps be using the Instant type instead since it’s pretty similar to the Moment object they were using anyway.

Btw, this author has an excellent summary on usage of ZonedDateTime and some of the ways that various platforms represent it’s string format since there currently isn’t a standard. I found it fascinating. :) https://stackoverflow.com/a/42196445

# In Summary

* tell developers to use `Instant` instead of ZonedDateTime if you’re okay with losing the zone information and if you want them to have an API that is an easier transition from MomentJS, Luzon, DayJS, or the actual JS date object.
* lastly, you guys are awesome. Thanks @pithu and @phueperz :)

**Comment by dgreene1** [comment]
I updated my comments above because I revisited this problem recently and I realized... _why would you wan to lose offset information?_

So my previous advice was bad because `Instant` converts to GMT and therefore you lose the offset information, which is a bummer since the service or the UI might want that offset information later when retrieving the data from the database.

Here's an example of what I'm trying to communicate:

```ts
const zdt = ZonedDateTime.now()

zdt.toOffsetDateTime().toString();
// '2022-04-28T11:04:35.853-04:00'

zdt.toInstant().toString()
// '2022-04-28T15:04:35.853Z'
```

Therefore, my original request for this issue can be answered by "no, please don't ask for a `.toISOString` but instead just use `.toOffsetDateTime().toString()`.
